### PR TITLE
Goto current line when opening internal text editor

### PR DESCRIPTION
**Pull request by gerhardol**
https://github.com/gitextensions/gitextensions/pull/11183#issuecomment-1703916102

## Proposed changes

Goto currently selected line in diff/tree/commit editor, when opening the internal editor (e.g. with F4).

## Screenshots 

image

### Before

image

### After

image

## Test methodology 

test added

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/Editor/FileViewer.cs [hunk 1]
```diff
@@ -406,6 +406,11 @@ public int GetLineFromVisualPosY(int visualPosY)
             return internalFileViewer.GetLineFromVisualPosY(visualPosY);
         }
 
+        public int? InitalLineNumber
+        {
+            set => internalFileViewer.InitalLineNumber = value;
```
Comment (RussKie) [inline GitUI/Editor/FileViewer.cs:411]: Set-only properties should be turned into methods.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 2]
```diff
@@ -2717,8 +2717,8 @@ private void editFileToolStripMenuItem_Click(object sender, EventArgs e)
                 return;
             }
 
-            UICommands.StartFileEditorDialog(fileName);
-
+            int? lineNumber = SelectedDiff.CurrentFileLine;
+            UICommands.StartFileEditorDialog(fileName, lineNumber: lineNumber);
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormCommit.cs:2721]: ❔ no need for a variable in this case
```suggestion
            UICommands.StartFileEditorDialog(fileName, lineNumber: SelectedDiff.CurrentFileLine);
```

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 3]
```diff
@@ -42,6 +42,7 @@ public FormEditor(GitUICommands commands, string? fileName, bool showWarning, bo
             panelMessage.Visible = showWarning;
 
             fileViewer.IsReadOnly = readOnly;
+            fileViewer.InitalLineNumber = lineNumber;
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormEditor.cs:45]: Why not just forward `lineNumber` to `OpenFile`?

File: GitUI/Editor/FileViewer.cs (hunk 1 above)
Comment (mstv) [reply GitUI/Editor/FileViewer.cs:411]: I would turn it into another argument to `ViewFileAsync` (advantage: stateless).

BTW: All calls of `ViewFileAsync` should be awaited or "filed".

File: GitUI/CommandsDialogs/FormEditor.cs (hunk 3 above)
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormEditor.cs]: The text is loaded in ShowDialog(), calling SetText(), so it cannot be set yet.

File: GitUI/CommandsDialogs/RevisionDiffControl.cs [hunk 4]
```diff
@@ -993,8 +993,8 @@ private void diffEditWorkingDirectoryFileToolStripMenuItem_Click(object sender,
                 return;
             }
 
-            var fileName = _fullPathResolver.Resolve(DiffFiles.SelectedItem.Item.Name);
-            UICommands.StartFileEditorDialog(fileName);
+            string? fileName = _fullPathResolver.Resolve(DiffFiles.SelectedItem.Item.Name);
+            UICommands.StartFileEditorDialog(fileName, lineNumber: DiffText.CurrentFileLine);
```
Comment (mstv) [inline GitUI/CommandsDialogs/RevisionDiffControl.cs:997]: needed here, too:
```suggestion
            int? lineNumber = BlameControl.Visible ? BlameControl.CurrentFileLine : DiffText.CurrentFileLine;
            UICommands.StartFileEditorDialog(fileName, lineNumber: lineNumber);
```

Comment (gerhardol) [reply GitUI/CommandsDialogs/RevisionDiffControl.cs]: Thanks!

**Review by mstv — APPROVED** [review]
👍

**Review by RussKie — APPROVED** [review]

**Pull request merged**